File: backend/src/modules/redemption-codes/redemption-code.service.ts
import crypto from 'node:crypto'

import { db } from '../../db/client.js'
import { AppError } from '../../types/app-error.js'
import type { RedemptionCodeRecord, RedemptionCodeType, SubscriptionRecord } from '../../types/models.js'
import { logActivity } from '../activity/activity.repository.js'
import { createSubscription } from '../subscriptions/subscription.repository.js'
import { createRedemptionCode, findCodeByCode, listCodes, markCodeUsed } from './redemption-code.repository.js'
import { generateCodesSchema, redeemCodeSchema } from './redemption-code.schema.js'

const typeToDays: Record<RedemptionCodeType, number> = {
  '1week': 7,
  '1month': 30,
  '3months': 90,
  '1year': 365,
}

export const generateCodes = (input: unknown, createdBy: number): RedemptionCodeRecord[] => {
  const parsed = generateCodesSchema.parse(input)
  const codes: RedemptionCodeRecord[] = []

  for (let i = 0; i < parsed.count; i++) {
    const code = crypto.randomBytes(16).toString('hex')
    const expiresAt = new Date(Date.now() + typeToDays[parsed.type] * 86400000).toISOString()
    const record = createRedemptionCode({
      code,
      type: parsed.type,
      createdBy,
      expiresAt,
    })
    codes.push(record)
  }

  return codes
}

export const redeemCode = (input: unknown, userId: number): SubscriptionRecord => {
  const parsed = redeemCodeSchema.parse(input)
  const code = findCodeByCode(parsed.code)

  if (!code || code.status !== 'unused') {
    throw new AppError('兑换码无效或已使用', 400)
  }

  if (new Date(code.expiresAt) < new Date()) {
    throw new AppError('兑换码已过期', 400)
  }

  const now = new Date()
  const endsAt = new Date(now.getTime() + typeToDays[code.type] * 86400000)

  const subscription = db.transaction(() => {
    markCodeUsed(code.id, userId)
    const sub = createSubscription({
      userId,
      startsAt: now.toISOString(),
      endsAt: endsAt.toISOString(),
      sourceCodeId: code.id,
    })
    logActivity(userId, 'redeem_code')
    return sub
  })()

  return subscription
}

export const getCodes = (): RedemptionCodeRecord[] => {
  return listCodes()
}

File: backend/app/services/redemption_code_service.py
import os
from datetime import datetime, timedelta, timezone

from ..database import get_db
from ..exceptions import AppError
from ..models import RedemptionCodeRecord, RedemptionCodeType, SubscriptionRecord
from ..repositories.activity_repo import log_activity
from ..repositories.redemption_code_repo import (
    create_redemption_code,
    find_code_by_code,
    list_codes,
    mark_code_used,
)
from ..repositories.subscription_repo import create_subscription

_TYPE_TO_DAYS: dict[str, int] = {
    "1week": 7,
    "1month": 30,
    "3months": 90,
    "1year": 365,
}


def generate_codes(code_type: RedemptionCodeType, count: int, created_by: int) -> list[RedemptionCodeRecord]:
    codes: list[RedemptionCodeRecord] = []
    for _ in range(count):
        code = os.urandom(16).hex()
        days = _TYPE_TO_DAYS[code_type]
        expires_at = (datetime.now(timezone.utc) + timedelta(days=days)).isoformat()
        record = create_redemption_code(
            code=code, code_type=code_type, created_by=created_by, expires_at=expires_at
        )
        codes.append(record)
    return codes


def redeem_code(code_str: str, user_id: int) -> SubscriptionRecord:
    code = find_code_by_code(code_str)
    if not code or code.status != "unused":
        raise AppError("兑换码无效或已使用", 400)

    if datetime.fromisoformat(code.expires_at) < datetime.now(timezone.utc):
        raise AppError("兑换码已过期", 400)

    now = datetime.now(timezone.utc)
    days = _TYPE_TO_DAYS[code.type]
    ends_at = now + timedelta(days=days)

    # Transaction: mark code + create subscription + log activity
    mark_code_used(code.id, user_id)
    subscription = create_subscription(
        user_id=user_id,
        starts_at=now.isoformat(),
        ends_at=ends_at.isoformat(),
        source_code_id=code.id,
    )
    log_activity(user_id, "redeem_code")

    return subscription


def get_codes() -> list[RedemptionCodeRecord]:
    return list_codes()
